File: converter/golden_parser.py
import xml.etree.ElementTree as ET
from dataclasses import dataclass, field
from typing import List, Optional


@dataclass
class PositionMark:
    name: str
    mark_type: int
    start: float
    num: int
    red: int
    green: int
    blue: int
    end: Optional[float] = None


@dataclass
class Tempo:
    inizio: float
    bpm: float
    metro: str
    battito: int


@dataclass
class Track:
    track_id: int
    name: str
    artist: Optional[str]
    album: Optional[str]
    genre: Optional[str]
    kind: str
    location: str
    size: int
    total_time: float
    track_number: Optional[int]
    year: Optional[int]
    average_bpm: float
    bit_rate: Optional[int]
    comments: Optional[str]
    tonality: Optional[str]
    label: Optional[str]
    sample_rate: int
    position_marks: List[PositionMark] = field(default_factory=list)
    tempo: Optional[Tempo] = None


@dataclass
class Playlist:
    name: str
    entries: int
    track_keys: List[int] = field(default_factory=list)


@dataclass
class GoldenModel:
    product_name: str
    product_version: str
    product_company: str
    collection_entries: int
    tracks: List[Track] = field(default_factory=list)
    playlists: List[Playlist] = field(default_factory=list)


def parse_position_mark(elem: ET.Element) -> PositionMark:
    end = None
    if "End" in elem.attrib:
        end = float(elem.attrib["End"])
    return PositionMark(
        name=elem.attrib["Name"],
        mark_type=int(elem.attrib["Type"]),
        start=float(elem.attrib["Start"]),
        num=int(elem.attrib["Num"]),
        red=int(elem.attrib["Red"]),
        green=int(elem.attrib["Green"]),
        blue=int(elem.attrib["Blue"]),
        end=end,
    )


def parse_tempo(elem: ET.Element) -> Tempo:
    return Tempo(
        inizio=float(elem.attrib["Inizio"]),
        bpm=float(elem.attrib["Bpm"]),
        metro=elem.attrib["Metro"],
        battito=int(elem.attrib["Battito"]),
    )


def parse_track(elem: ET.Element) -> Track:
    position_marks = []
    tempo = None
    for child in elem:
        if child.tag == "POSITION_MARK":
            position_marks.append(parse_position_mark(child))
        elif child.tag == "TEMPO":
            tempo = parse_tempo(child)

    def get_optional_int(name: str) -> Optional[int]:
        val = elem.attrib.get(name)
        return int(val) if val is not None else None

    return Track(
        track_id=int(elem.attrib["TrackID"]),
        name=elem.attrib["Name"],
        artist=elem.attrib.get("Artist"),
        album=elem.attrib.get("Album"),
        genre=elem.attrib.get("Genre"),
        kind=elem.attrib["Kind"],
        location=elem.attrib["Location"],
        size=int(elem.attrib["Size"]),
        total_time=float(elem.attrib["TotalTime"]),
        track_number=get_optional_int("TrackNumber"),
        year=get_optional_int("Year"),
        average_bpm=float(elem.attrib["AverageBpm"]),
        bit_rate=get_optional_int("BitRate"),
        comments=elem.attrib.get("Comments"),
        tonality=elem.attrib.get("Tonality"),
        label=elem.attrib.get("Label"),
        sample_rate=int(elem.attrib["SampleRate"]),
        position_marks=position_marks,
        tempo=tempo,
    )


def parse_playlist(elem: ET.Element) -> Playlist:
    track_keys = [
        int(child.attrib["Key"]) for child in elem if child.tag == "TRACK"
    ]
    return Playlist(
        name=elem.attrib["Name"],
        entries=int(elem.attrib["Entries"]),
        track_keys=track_keys,
    )


def parse_golden_xml(path: str) -> GoldenModel:
    tree = ET.parse(path)
    root = tree.getroot()

    product = root.find("PRODUCT")
    product_name = product.attrib["Name"]
    product_version = product.attrib["Version"]
    product_company = product.attrib["Company"]

    collection = root.find("COLLECTION")
    collection_entries = int(collection.attrib["Entries"])

    tracks = [parse_track(track_elem) for track_elem in collection.findall("TRACK")]

    playlists = []
    playlists_elem = root.find("PLAYLISTS")
    if playlists_elem is not None:
        root_node = playlists_elem.find("NODE")
        if root_node is not None:
            for node in root_node.findall("NODE"):
                if node.attrib.get("Type") == "1":
                    playlists.append(parse_playlist(node))

    return GoldenModel(
        product_name=product_name,
        product_version=product_version,
        product_company=product_company,
        collection_entries=collection_entries,
        tracks=tracks,
        playlists=playlists,
    )

File: converter/golden_parser.js
import { readFileSync } from 'node:fs';
import { XMLParser } from 'fast-xml-parser';

export class PositionMark {
  constructor({ name, mark_type, start, num, red, green, blue, end = null }) {
    this.name = name;
    this.mark_type = mark_type;
    this.start = start;
    this.num = num;
    this.red = red;
    this.green = green;
    this.blue = blue;
    this.end = end;
  }
}

export class Tempo {
  constructor({ inizio, bpm, metro, battito }) {
    this.inizio = inizio;
    this.bpm = bpm;
    this.metro = metro;
    this.battito = battito;
  }
}

export class Track {
  constructor({
    track_id, name, artist = null, album = null, genre = null,
    kind, location, size, total_time, track_number = null,
    year = null, average_bpm, bit_rate = null, comments = null,
    tonality = null, label = null, sample_rate,
    position_marks = [], tempo = null,
  }) {
    this.track_id = track_id;
    this.name = name;
    this.artist = artist;
    this.album = album;
    this.genre = genre;
    this.kind = kind;
    this.location = location;
    this.size = size;
    this.total_time = total_time;
    this.track_number = track_number;
    this.year = year;
    this.average_bpm = average_bpm;
    this.bit_rate = bit_rate;
    this.comments = comments;
    this.tonality = tonality;
    this.label = label;
    this.sample_rate = sample_rate;
    this.position_marks = position_marks;
    this.tempo = tempo;
  }
}

export class Playlist {
  constructor({ name, entries, track_keys = [] }) {
    this.name = name;
    this.entries = entries;
    this.track_keys = track_keys;
  }
}

export class GoldenModel {
  constructor({
    product_name, product_version, product_company,
    collection_entries, tracks = [], playlists = [],
  }) {
    this.product_name = product_name;
    this.product_version = product_version;
    this.product_company = product_company;
    this.collection_entries = collection_entries;
    this.tracks = tracks;
    this.playlists = playlists;
  }
}

function parsePositionMark(elem) {
  return new PositionMark({
    name: elem['@_Name'],
    mark_type: parseInt(elem['@_Type'], 10),
    start: parseFloat(elem['@_Start']),
    num: parseInt(elem['@_Num'], 10),
    red: parseInt(elem['@_Red'], 10),
    green: parseInt(elem['@_Green'], 10),
    blue: parseInt(elem['@_Blue'], 10),
    end: elem['@_End'] !== undefined ? parseFloat(elem['@_End']) : null,
  });
}

function parseTempo(elem) {
  return new Tempo({
    inizio: parseFloat(elem['@_Inizio']),
    bpm: parseFloat(elem['@_Bpm']),
    metro: elem['@_Metro'],
    battito: parseInt(elem['@_Battito'], 10),
  });
}

function parseIntOr(val) {
  return val !== undefined ? parseInt(val, 10) : null;
}

function parseTrack(elem) {
  const positionMarks = [];
  let tempo = null;

  const pos = elem.POSITION_MARK;
  if (pos) {
    const arr = Array.isArray(pos) ? pos : [pos];
    for (const p of arr) {
      positionMarks.push(parsePositionMark(p));
    }
  }

  const te = elem.TEMPO;
  if (te) {
    const arr = Array.isArray(te) ? te : [te];
    tempo = parseTempo(arr[0]);
  }

  return new Track({
    track_id: parseInt(elem['@_TrackID'], 10),
    name: elem['@_Name'],
    artist: elem['@_Artist'] || null,
    album: elem['@_Album'] || null,
    genre: elem['@_Genre'] || null,
    kind: elem['@_Kind'],
    location: elem['@_Location'],
    size: parseInt(elem['@_Size'], 10),
    total_time: parseFloat(elem['@_TotalTime']),
    track_number: parseIntOr(elem['@_TrackNumber']),
    year: parseIntOr(elem['@_Year']),
    average_bpm: parseFloat(elem['@_AverageBpm']),
    bit_rate: parseIntOr(elem['@_BitRate']),
    comments: elem['@_Comments'] || null,
    tonality: elem['@_Tonality'] || null,
    label: elem['@_Label'] || null,
    sample_rate: parseInt(elem['@_SampleRate'], 10),
    position_marks: positionMarks,
    tempo,
  });
}

function parsePlaylist(elem) {
  const trackKeys = [];
  if (elem.TRACK) {
    const arr = Array.isArray(elem.TRACK) ? elem.TRACK : [elem.TRACK];
    for (const t of arr) {
      trackKeys.push(parseInt(t['@_Key'], 10));
    }
  }
  return new Playlist({
    name: elem['@_Name'],
    entries: parseInt(elem['@_Entries'], 10),
    track_keys: trackKeys,
  });
}

export function parseGoldenXml(path) {
  const xml = readFileSync(path, 'utf-8');

  const parser = new XMLParser({
    ignoreAttributes: false,
    attributeNamePrefix: '@_',
    allowBooleanAttributes: true,
    ignoreDeclaration: true,
    isArray: (name) =>
      name === 'TRACK' || name === 'POSITION_MARK' || name === 'NODE' || name === 'TEMPO',
  });

  const doc = parser.parse(xml);
  const root = doc.DJ_PLAYLISTS;

  const product = root.PRODUCT;
  const productName = product['@_Name'];
  const productVersion = product['@_Version'];
  const productCompany = product['@_Company'];

  const collection = root.COLLECTION;
  const collectionEntries = parseInt(collection['@_Entries'], 10);

  const rawTracks = collection.TRACK || [];
  const tracks = rawTracks.map(parseTrack);

  const playlists = [];
  const playlistsElem = root.PLAYLISTS;
  if (playlistsElem) {
    const rootNodeArr = playlistsElem.NODE;
    if (rootNodeArr) {
      const nodes = Array.isArray(rootNodeArr) ? rootNodeArr : [rootNodeArr];
      for (const node of nodes) {
        if (Array.isArray(node.NODE)) {
          for (const child of node.NODE) {
            if (child['@_Type'] === '1') {
              playlists.push(parsePlaylist(child));
            }
          }
        }
      }
    }
  }

  return new GoldenModel({
    product_name: productName,
    product_version: productVersion,
    product_company: productCompany,
    collection_entries: collectionEntries,
    tracks,
    playlists,
  });
}
